### PR TITLE
Structural/ Adding the Serial-Parallel rule of mixtures

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm adding to the StrucMech App the Serial-Parallel rule of mixtures developed in the F. Rastellini thesis.
In order to implement this I've used all the implementations of the Subproperties, so it is included in this PR unless the SubProps PR is merged

### CODE REVIEW

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Blocking until https://github.com/KratosMultiphysics/Kratos/pull/2919 is merged

Reviewing now

**Comment by AlejandroCornejo** [comment]
> Blocking until #2919 is merged
> 
> Reviewing now

image

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: This should depend of the input laws!!!

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Add size

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Technically std::size_t

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Add sizes

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: noalias()

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Size

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: noalias()

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Tabulation is missing

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Tabulation

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: std::size_t

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Tabulation

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Size

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Remove it

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Remove it or call the elastic law

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Size

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Size

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.h
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.h]: You add here the link to the thesis

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.h
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.h]: Wrong, you base law is ConstitutiveLaw

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.h
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.h]: Add the Doxygen

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Here some comments, please add tests

**Comment by philbucher** [comment]
@AlejandroCornejo can you please add `[skip ci]` to your commit-messages while this is under development?
This way you won't block the CI
Thanks!

**Comment by AlejandroCornejo** [comment]
> @AlejandroCornejo can you please add `[skip ci]` to your commit-messages while this is under development?
> This way you won't block the CI
> Thanks!

Ok!! I have to write [skip ci] before the commit message. Sure

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: this formulation is developed in 3D... Another problem is that here the input laws do not exist right?

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: I would use the information from the input laws

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: done

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: done in all cases

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: done

File: applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: done

**Comment by frastellini** [comment]
Thank you @AlejandroCornejo for the implementation and @loumalouomega for the kind corrections.
The implementation looks very accurate.
Even-though the general implementation includes the projector tensor, I do not recommend its use since it implies a lot of matrix multiplications. I think we can save a lot of computing time if we particularise the solution for the projector with fibre in local X direction. I strongly suggest the implementation of the particular case (fibre in local X direction) if we want improve computing time performance.
So that, we will be able to use the general case with projector tensor, and also the quicker version for unidirectional composites where fibre is considered parallel to local material x-axis.

**Comment by AlejandroCornejo** [comment]
ping @RiccardoRossi

**Comment by AlejandroCornejo** [comment]
> Thank you @AlejandroCornejo for the implementation and @loumalouomega for the kind corrections.
> The implementation looks very accurate.
> Even-though the general implementation includes the projector tensor, I do not recommend its use since it implies a lot of matrix multiplications. I think we can save a lot of computing time if we particularise the solution for the projector with fibre in local X direction. I strongly suggest the implementation of the particular case (fibre in local X direction) if we want improve computing time performance.
> So that, we will be able to use the general case with projector tensor, and also the quicker version for unidirectional composites where fibre is considered parallel to local material x-axis.

Thank you @frastellini, I wanted to implement this general version just in case that someone wants to simulate biparallel structures (fibers in 2 axis like in 2 way floor slabs) and so on.
I suppose that someone can implement a derived version with the minor changes requested for this simpler formulation.
Currently I have to attend other implementations :)

**Review by frastellini — COMMENTED** [review]
Yes. I understand.
The implementation has the approval from my side.
Anyway, we will need to wait until the PR #2919 (sub-properties) is approved first.

**Review by frastellini — DISMISSED** [review]
Good work!

**Comment by loumalouomega** [comment]
> Good work!

I will remove my block as soon as the subproperties are merged. Note that I could be evil and remove my block, then the subproperties would be merged.

**Comment by loumalouomega** [comment]
@AlejandroCornejo I solved a conflict, check I haven't broke anything

**Comment by AlejandroCornejo** [comment]
How it's the state of this pr??

**Comment by loumalouomega** [comment]
> How it's the state of this pr??



Depends on #2919. Besides, Travis and appveyor are failing

**Comment by AlejandroCornejo** [comment]
but it should be merged after the Release right? @loumalouomega @RiccardoRossi

**Comment by philbucher** [comment]
@AlejandroCornejo this PR depends on the SubProperties which is a major feature to be added to the Core and hence has to be reviewed and agreed on by the @KratosMultiphysics/technical-committee first, which unfortunately takes time :/

**Comment by loumalouomega** [comment]
> but it should be merged after the Release right? @loumalouomega @RiccardoRossi

You can fix Appveyor and Travis compilation meanwhile :P

**Comment by AlejandroCornejo** [comment]
> > but it should be merged after the Release right? @loumalouomega @RiccardoRossi
> 
> You can fix Appveyor and Travis compilation meanwhile :P

It seems that there's no definition of the p_prop in the line 181 in the read_materials_utility.cpp but I do not want to touch anything because I can mess everything I think

**Comment by loumalouomega** [comment]
@AlejandroCornejo Check the current implementation on #2919, I correct the conflicts regularly (must be the same or there will be a conflict)

**Comment by AlejandroCornejo** [comment]
> @AlejandroCornejo Check the current implementation on #2919, I correct the conflicts regularly (must be the same or there will be a conflict)

luckily I've done the same jeje

**Comment by loumalouomega** [comment]
@AlejandroCornejo conflict!!

**Comment by AlejandroCornejo** [comment]
> @AlejandroCornejo conflict!!

I know! I want to wait until the other branch is merged and then I'll merge and solve everything

**Comment by philbucher** [comment]
please move the law to the advanced Claws folder

**Comment by AlejandroCornejo** [comment]
> 
> 
> please move the law to the advanced Claws folder

yes sure, I'm updating everything, I need more time ;)

**Comment by AlejandroCornejo** [comment]
Now everything works again and moved to the advanced CL :)

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Technically this are brand new, not references, you can remove the & and r_

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Idem for the followings

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Please do not hardcode this, define a variable and add it to properties. take this value in case the property is not defined

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: spaces are worng

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: check spaces in this method

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: spaces

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp:469]: Do you need this?, don't you prefer to call the linear elastic law?

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Capital
```suggestion
        // Total strain vector
```

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: Spaces

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.h
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.h:466]: Document the private and protected methods

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Great work, check my comments

File: applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/custom_advanced_constitutive/serial_parallel_rule_of_mixtures_law.cpp]: done

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: how long does this test need? This is the small suite, it should not take more than ~0.05s max

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: It is exactly the same as the previous ones (TSimpleSmallDeformationPlasticityVMTest) but changing the CL to this new composite rule of mixtures :)

**Comment by AlejandroCornejo** [comment]
I think I've adderessed your comments @loumalouomega :) thank you all

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: and how long does that one run? :)

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: let me check

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: Maybe we should move some of the tests to Nightly, including the old ones.

**Comment by loumalouomega** [comment]
> 
> 
> I think I've adderessed your comments @loumalouomega :) thank you all

I commented and agreed to move the tests to the Nightly suite

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: In another PR then right?

Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: I think it takes like 0.5s so maybe it's too long

**Comment by loumalouomega** [comment]
Think your computer is way faster than the computer runs the tests, must be
even lowert

File: applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py
Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/test_StructuralMechanicsApplication.py]: yep that shoudl go to the night suite

**Review by loumalouomega — APPROVED** [review]
Finally

**Comment by AlejandroCornejo** [comment]
> 
> 
> Finally

7 years later, in 1999, it's still kicking :) @loumalouomega

**Pull request merged**
